[ENHANCEMENT] - Add custom function support

Glenn-Ewing opened the pull request:
**Change Description:**
This PR implements changes which allow for custom functions by Exo migrations. (create|replace|drop)

The definition of FunctionMigrations is a bit more involved than views, but simpler than standard Migrations (tables) in terms of their classification. The approach and intent are that a custom function is immutable, and can only be created, replaced, or dropped (not modified).

Since we have diverged from Migrations for views, and  Functions also have their own operations such as Parameters, ReturnType, Variables, etc. I (again) did not try to make Migrations work for functions and opted for the new class.

As with views, since there is not a 'simple' way to manage the body of a function, the body of a function relies on passing a string, that said, management of parameter and variable data types should be simplified and consistent due to the Exo Operations added.

**Changes made:**
- Create FunctionMigration Class (Parallel to Migration & ViewMigration)
- Create FunctionOperation Class (Parallel to TableOperation)
- Create Sub Operation Classes (ReturnTypeOperation, ParameterOperation & VariableOperation)

- Update StatementHandler implementations to create the appropriate SQL based on class type, etc.
- Update History and Handlers to handle the classes by type.
- Add/update test cases to validate changes as needed.

**Example migration:**

```
<?php

return Exo\FunctionMigration::create('my_function')
            ->isDeterministic(true)
            ->withReturnType('integer')
            ->addParameter('someParameter', ['type' => 'integer'])
            ->addVariable('someVariable', ['type' => 'string', 'length' => 64])
            ->withBody('RETURN someParameter * 10;');
```

**Reviewer Notes**
Please review the code for sanity checks, run tests, and put FunctionMigrations through their paces to your heart's content.

FYI Re: the SQL used to create the function:
I am not a huge fan of the implementation for the Create/replace script, which simply does a:
```
DROP IF EXISTS {functionName};
CREATE FUNCTION {...}
```
But CREATE OR REPLACE was not working, so this seemed like a pragmatic solution at least for now.

Let me know if there is a preferred approach to this type of enhancement, I am happy to refactor if needed.

Glenn-Ewing committed:
WIP Begin implementing function operations

Glenn-Ewing committed:
Merge branch 'enhancement-add-view-support' into enhancement-add-custom-function-support

Glenn-Ewing committed:
Update Tests for Function Support

Glenn-Ewing committed:
Rename Test for Accuracy: ViewTest -> ViewMigrationTest

Glenn-Ewing committed:
Add Function Migration and Operation Classes

Glenn-Ewing committed:
Update History & Statement Builder to support Function Operations

joshmcrae commented on src/FunctionMigration.php:
```diff
@@ -0,0 +1,248 @@
+    }
+
+    /**
+     * Migration constructor.
```
This should match the class name.

joshmcrae commented on src/FunctionMigration.php:
```diff
@@ -0,0 +1,248 @@
+     */
+    public static function create(string $name)
+    {
+        return new self($name, FunctionOperation::CREATE);
```
Just as a matter of consistency, in other static class methods we've type-hinted the return type according to the class and constructed the instance using `new ClassName`. Of course this implies that we're not expecting the classes to be extended so either we update the existing instances of mark the classes as `final`.

Example [here](https://github.com/tithely/exo/blob/master/src/Migration.php). Keen to hear your feedback on this.

Glenn-Ewing replied:
Right on, I see your point totally. I don't think I see a need for extending these at this point, and we can always refactor later if needed. I will go ahead and set the returns to the concrete class and set the migration classes as final across the board.

joshmcrae commented on src/History.php:
```diff
@@ -142,13 +143,16 @@ private static function getEntityName(AbstractOperation $operation) {
 
         switch($operationClass) {
-
+            // TODO: Refactor all to use getName() across the board?
```
That's a good idea given that they all descend from `AbstractOperation`. It should probably be an abstract method.

Glenn-Ewing replied:
Done as follows:
1. Created an Operation Interface for type hinting
2. Created the abstract class implement the interface and a concrete protected implementation of getName()
3. Refactored all migration classes away from getTable(), getView(), etc. to getName()

joshmcrae commented on src/Statement/MysqlStatementBuilder.php:
```diff
@@ -4,13 +4,32 @@
 {
+    const VIEW_CREATE_OR_REPLACE = 'CREATE OR REPLACE VIEW %s AS (%s);';
+    const VIEW_DROP = 'DROP VIEW %s;';
+    const FUNCTION_CREATE_OR_REPLACE = '
```
Is there a need for idempotency when creating functions? I would expect this operation to fail if the functional already exists, just as I would expect the operation to fail when I try to create an existing table.

Glenn-Ewing replied:
I don't know that there is an explicit need for it, it just seemed the simpler solution for items that are self encapsulated. I am open to change this, but considering the entire state of the view and or function is encapsulated in the definition, I saw no risk as well as convenience & simplicity. If you're ok with it, I'd like to go with this for now and swing back to update this in the near term. I say this mostly because I need to implement views and functions in Messaging and am trying to take the straightest approach to that for the moment. Let me know if you think this will work for now.

joshmcrae replied:
For consistency I think it's best that create and alter result in different SQL being issued, such that create fails if the function/view already exists.

Glenn-Ewing replied:
Done.

joshmcrae reviewed (changes requested):
This is excellent! I've got some small issues to resolve with consistency and a question about function creation.

Glenn-Ewing committed:
Minor Refactor to Inheritence and Interfaces (Part 1: Tests)

Update tests to reflect intent to use getName() as a universal operation method to return the name of the intended operation entity (table name, column name, view name, etc.) since the type of the entity is obvious from the parent class name, etc.

Glenn-Ewing committed:
Change Abstract Operation to Operation Interface

There were no concrete implementations of methods, etc. so the use of an interface seemed more appropriate for type hinting, etc.

Glenn-Ewing committed:
Minor Refactor to Inheritence and Interfaces (Part 2: Implementation)

Udpate all operation references to use getName() vs. getTable(), getView(), etc..

Glenn-Ewing committed:
Minor Refactor to Inheritence and Interfaces (Part 2: Implemetation continued)

Remove unecessary helper methods to favor use of getName().

Glenn-Ewing committed:
Update classes to use OperationInterface vs. AbstractOperation

Glenn-Ewing committed:
Update Migrations

1. Make Migrations final classes
2. Make methods return the current class by type vs 'self'

Glenn-Ewing committed:
Update usages

Glenn-Ewing committed:
Abstract getName() out of Operations

Glenn-Ewing commented:
@joshmcrae - Take a peek, I've made all the changes you suggested.  See replies for specifics to the respective comments.

joshmcrae commented on src/Operation/OperationInterface.php:
```diff
@@ -0,0 +1,20 @@
+
+namespace Exo\Operation;
+
+interface OperationInterface
```
Could we get by without the interface, simply type-hinting against the abstract class?

Glenn-Ewing replied:
Sure! Updates made to reflect this change.

joshmcrae committed:
Fixed test cases

joshmcrae committed:
Updated Handler::migrate() and Handler::rollback() signatures

Also improved formatting of History class

joshmcrae committed:
Implemented History::clone()

joshmcrae committed:
Implemented support for array of executed versions in Handler::migrate()

joshmcrae committed:
Implemented support for array of executed versions in Handler::rollback()

joshmcrae committed:
Cleaned up Handler::rollback() and test case

Glenn-Ewing committed:
Merge branch 'enhance-unordered-history' into enhancement-add-custom-function-support

Glenn-Ewing committed:
Update MysqlStatementBuilderTest.php

Update tests to reflect updates to SQL usages.

Glenn-Ewing committed:
Update AbstractOperation.php

Remove OperationInterface

Glenn-Ewing committed:
Update MysqlStatementBuilder.php

Remove Operation Interface

Glenn-Ewing committed:
Update MysqlStatementBuilder.php

Update SQL Usages, removing idempotency

Glenn-Ewing committed:
Update MysqlHandlerTest.php

Update tests to include the presence of a function migration.

Glenn-Ewing committed:
Misc Cleanup

Glenn-Ewing commented:
@joshmcrae all changes made, please review as your earliest convenience.

Glenn-Ewing committed:
Update function tests and syntax for parameters

Glenn-Ewing committed:
Syntax updates

Glenn-Ewing reviewed (commented):
Self-review (post updates) completed.

joshmcrae committed:
Merge branch 'master' into enhancement-add-custom-function-support

joshmcrae commented:
Released in [0.2.0](https://github.com/tithely/exo/releases/tag/0.2.0).
